planner/memo/group.go

```diff
@@ -21,6 +21,55 @@ import (
 	"github.com/pingcap/tidb/planner/property"
 )
 
+// EngineType is determined by whether it's above or below `Gather`s.
+// Plan will choose the different engine to be implemented/executed on according to its EngineType.
+// Different engine may support different operators with different cost, so we should design
+// different transformation and implementation rules for each engine.
+type EngineType uint
+
+const (
+	// EngineTiDB stands for groups which is above `Gather`s and will be executed in TiDB layer.
+	EngineTiDB EngineType = 1 << iota
+	// EngineTiKV stands for groups which is below `Gather`s and will be executed in TiKV layer.
+	EngineTiKV
+	// EngineTiFlash stands for groups which is below `Gather`s and will be executed in TiFlash layer.
+	EngineTiFlash
+)
+
+// EngineTypeSet is the bit set of EngineTypes.
+type EngineTypeSet uint
+
+const (
+	// EngineTiDBOnly is the EngineTypeSet for EngineTiDB only.
+	EngineTiDBOnly = EngineTypeSet(EngineTiDB)
+	// EngineTiKVOnly is the EngineTypeSet for EngineTiKV only.
+	EngineTiKVOnly = EngineTypeSet(EngineTiKV)
+	// EngineTiFlashOnly is the EngineTypeSet for EngineTiFlash only.
+	EngineTiFlashOnly = EngineTypeSet(EngineTiFlash)
+	// EngineTiKVOrTiFlash is the EngineTypeSet for (EngineTiKV | EngineTiFlash).
+	EngineTiKVOrTiFlash = EngineTypeSet(EngineTiKV | EngineTiFlash)
+	// EngineAll is the EngineTypeSet for all of the EngineTypes.
+	EngineAll = EngineTypeSet(EngineTiDB | EngineTiKV | EngineTiFlash)
+)
+
+// Contain checks whether the EngineTypeSet contains the EngineType.
+func (e EngineTypeSet) Contain(tp EngineType) bool {
```

Comment:
How about `Contains`, like the function name in the [`strings` package](https://golang.org/pkg/strings/)
```suggestion
func (e EngineTypeSet) Contains(tp EngineType) bool {
```

Reply:
addressd